sw/host/opentitanlib/src/spiflash/flash.rs

```diff
@@ -0,0 +1,282 @@
+    pub const LEGACY_PAGE_SIZE: u32 = 256;
+
+    /// Status register bits:
+    /// The `WIP` bit indicates a write in progress (sometimes called the busy bit).
+    pub const STATUS_WIP: u8 = 0x01;
+    /// The `WEL` bit is the write enable latch.
+    pub const STATUS_WEL: u8 = 0x02;
+
+    // Prepare an opcode and address buffer according to the address_mode.
+    fn opcode_with_address(&self, opcode: u8, address: u32) -> Result<Vec<u8>> {
+        if address < self.size {
+            let mut buf = Vec::new();
+            buf.push(opcode);
+            if self.address_mode == AddressMode::Mode4b {
+                buf.push((address >> 24) as u8);
+            }
+            buf.push((address >> 16) as u8);
+            buf.push((address >> 8) as u8);
+            buf.push((address >> 0) as u8);
+            Ok(buf)
+        } else {
+            Err(Error::AddressOutOfBounds(address, self.size).into())
+        }
+    }
+
+    /// Read `length` bytes of the JEDEC ID from the `spi` target.
+    pub fn read_jedec_id(spi: &dyn Target, length: usize) -> Result<Vec<u8>> {
+        let mut buf = vec![0u8; length];
+        spi.run_transaction(&mut [
+            Transfer::Write(&[SpiFlash::READ_ID]),
+            Transfer::Read(&mut buf),
+        ])?;
+        Ok(buf)
+    }
+
+    /// Read status register from the `spi` target.
+    pub fn read_status(spi: &dyn Target) -> Result<u8> {
+        let mut buf = [0u8; 1];
+        spi.run_transaction(&mut [
+            Transfer::Write(&[SpiFlash::READ_STATUS]),
+            Transfer::Read(&mut buf),
+        ])?;
+        Ok(buf[0])
+    }
+
+    /// Poll the status register waiting for the busy bit to clear.
+    pub fn wait_for_busy_clear(spi: &dyn Target) -> Result<()> {
+        while SpiFlash::read_status(spi)? & SpiFlash::STATUS_WIP != 0 {
+            // Do nothing.
+        }
+        Ok(())
+    }
+
+    /// Send the WRITE_ENABLE opcode to the `spi` target.
+    pub fn set_write_enable(spi: &dyn Target) -> Result<()> {
+        let wren = [SpiFlash::WRITE_ENABLE];
+        spi.run_transaction(&mut [Transfer::Write(&wren)])
+    }
+
+    /// Read and parse the SFDP table from the `spi` target.
+    pub fn read_sfdp(spi: &dyn Target) -> Result<Sfdp> {
+        let mut buf = vec![0u8; 256];
+        let mut tries = 0;
+        loop {
+            spi.run_transaction(&mut [
+                // READ_SFDP always takes a 3-byte address followed by a dummy
+                // byte regardless of address mode.  To simplify, we always
+                // read from address zero.
+                Transfer::Write(&[SpiFlash::READ_SFDP, 0, 0, 0, 0]),
+                Transfer::Read(&mut buf),
+            ])?;
+
+            // We only want to give SFDP parsing one extra chance for length
+            // extension. If parsing fails a second time, just return the error.
+            let result = Sfdp::try_from(&buf[..]);
+            if result.is_ok() || tries > 0 {
+                return Ok(result?);
```

Comment:
Wouldn't `result.into()` make more sense? Or does that not quite work?

Reply:
In order to do that, I need to make TryFrom for Sfdp have an Error type of anyhow::Error.

That's fine by me, as that's the expected way to convert from a byte buffer into an Sfdp and it will internally convert all of the other errors into anyhow::Error.